### PR TITLE
Speed up empty highlighting many fields

### PR DESCRIPTION
**Pull request by nik9000**
Kibana often highlights *everything* like this:
```
POST /_search
{
  "query": ...,
  "size": 500,
  "highlight": {
    "fields": {
      "*": { ... }
    }
  }
}
```

This can get slow when there are hundreds of mapped fields. I tested
this locally and unscientifically and it took a request from 20ms to
150ms when there are 100 fields. I've seen clusters with 2000 fields
where simple search go from 500ms to 1500ms just by turning on this sort
of highlighting. Even when the query is just a `range` that and the
fields are all numbers and stuff so it won't highlight anything.

This speeds up the `unified` highlighter in this case in a few ways:
1. Build the highlighting infrastructure once field rather than once pre
   document per field. This cuts out a *ton* of work analyzing the query
   over and over and over again.
2. Bail out of the highlighter before loading values if we can't produce
   any results.

Combined these take that local 150ms case down to 65ms. This is unlikely
to be really useful when there are only a few fetched docs and only a
few fields, but we often end up having many fields with many fetched
docs.

### CODE REVIEW

**Comment by nik9000** [comment]
We don't have much in the way of rally tests for highlighting. It'd be nice to have something though.

I'm waiting on a snapshot to do "more real" performance testing, but I think this is ready to look it.

**Comment by nik9000** [comment]
I'm sure there are *better* ways to do this, but I wanted to be fairly surgical here to try and get this in without a lot of refactoring.

**Comment by gingerwizard** [comment]
@nik9000 i think it makes sense to enhance the PMC track for this one. We're re-evaluating tracks but this one is a full text corpus and it should be fairly easy to add a highlighting query [here](https://github.com/elastic/rally-tracks/blob/master/pmc/operations/default.json). I'll raise a ticket in the track repo.

**Comment by nik9000** [comment]
@romseygeek it looks like this one is still useful. I still see some savings locally and am trying to get some better test data.

I think it'd be cleaner to implement this by refactoring all of the highlighters to build themselves in the "processor construction" phase. But this is "smaller".

**Comment by nik9000** [comment]
@elasticsearchmachine run elasticsearch-ci/2

File: plugins/mapper-annotated-text/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/AnnotatedTextHighlighter.java [hunk 1]
```diff
@@ -37,39 +38,37 @@
...
+        ((AnnotatedHighlighterAnalyzer) highlighter.getIndexAnalyzer()).setAnnotations(annotations);
+        return strings;
+    }
 
-        ArrayList<Object> result = new ArrayList<>(annotations.length);
-        for (int i = 0; i < annotations.length; i++) {
-            result.add(annotations[i].textMinusMarkup);
-        }
-        return result;
+    @Override
```
Comment (romseygeek) [inline plugins/mapper-annotated-text/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/AnnotatedTextHighlighter.java:64]: I think this change means we can remove `hitContext` from the method sig here as well - we only added it in to access the cache.

File: plugins/mapper-annotated-text/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/AnnotatedTextHighlighterTests.java [hunk 2]
```diff
@@ -93,17 +88,14 @@ private void assertHighlightOneDoc(String fieldName, String []markedUpInputs,
         IndexSearcher searcher = newSearcher(reader);
         iw.close();
 
-        LeafReaderContext context = searcher.getIndexReader().leaves().get(0);
-        HitContext mockHitContext = new HitContext(null, context, 0, null, new HashMap<>());
-        AnnotatedHighlighterAnalyzer hiliteAnalyzer = new AnnotatedHighlighterAnalyzer(wrapperAnalyzer, mockHitContext);
```
Comment (romseygeek) [inline plugins/mapper-annotated-text/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/AnnotatedTextHighlighterTests.java:98]: ❤️

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/UnifiedHighlighter.java [hunk 3]
```diff
@@ -60,84 +62,97 @@ public boolean canHighlight(MappedFieldType fieldType) {
...
-            if ((offsetSource == OffsetSource.ANALYSIS) && (fieldValueLength > maxAnalyzedOffset)) {
-                throw new IllegalArgumentException(
-                    "The length of [" + fieldContext.fieldName + "] field of [" + hitContext.hit().getId() +
-                        "] doc of [" + context.index().getName() + "] index " + "has exceeded [" +
-                        maxAnalyzedOffset + "] - maximum allowed to be analyzed for highlighting. " +
-                        "This maximum can be set by changing the [" + IndexSettings.MAX_ANALYZED_OFFSET_SETTING.getKey() +
-                        "] index level setting. " + "For large texts, indexing with offsets or term vectors is recommended!");
+        @SuppressWarnings("unchecked")
+        Map<String, CustomUnifiedHighlighter> cache = (Map<String, CustomUnifiedHighlighter>) fieldContext.hitContext.cache()
+            .computeIfAbsent(UnifiedHighlighter.class.getName(), k -> new HashMap<>());
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/UnifiedHighlighter.java:67]: Pretty sure we can rework the highlighter API now to remove the need for this dumb cache, which would be a really nice cleanup.

**Review by romseygeek — APPROVED** [review]
LGTM! Thanks for working through this, I'll take a look at re-jigging the highlighter API so that it fits better with the new FetchSubPhase API as well.

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/UnifiedHighlighter.java (hunk 3 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/UnifiedHighlighter.java:67]: It totally would be!

**Comment by nik9000** [comment]
Before:
```
curl -XPOST localhost:9200/_search -d'{
  "size": 500,
  "query": <something stolen from kibana's discover tab when there is just a date range>,
  "highlight": {
    "fields": {
      "*": { ... }
    }
  }
} | grep took | head -n 1
  "took" : 710,
```

After:
```
curl -XPOST localhost:9200/_search -d'{
  "size": 500,
  "query": <something stolen from kibana's discover tab when there is just a date range>,
  "highlight": {
    "fields": {
      "*": { ... }
    }
  }
} | grep took | head -n 1
  "took" : 300,
```

**Comment by nik9000** [comment]
Those before and after numbers are pretty unscientific and come from a snapshot of a cluster in use at elastic restored to my desktop. It is the one I described in the PR description who's searches jumped from 500ms to 1500ms just by adding noop highlighting. So I'm not sure what it'll do to that cluster. Hopefully it'll help!

Also:
```
curl -XPOST localhost:9200/_search -d'{
  "size": 500,
  "query": <something stolen from kibana's discover tab when there is just a date range>,
} | grep took | head -n 1
  "took" : 290,
```

**Pull request merged**